x-pack/plugins/security_solution/server/endpoint/services/actions/create/write_action_to_indices.ts

```diff
@@ -0,0 +1,214 @@
+    esClient,
+    logger,
+    dataStreamName: ENDPOINT_ACTIONS_DS,
+  });
+
+  // if the new endpoint indices/data streams exists
+  // write the action request to the new endpoint index
+  if (doesLogsEndpointActionsDsExist) {
+    const logsEndpointActionsResult = await esClient.index<LogsEndpointAction>(
+      {
+        index: ENDPOINT_ACTIONS_INDEX,
+        body: {
+          ...doc,
+          agent: {
+            id: payload.endpoint_ids,
+          },
+        },
+        refresh: 'wait_for',
+      },
+      { meta: true }
+    );
+    if (logsEndpointActionsResult.statusCode !== 201) {
+      throw new Error(logsEndpointActionsResult.body.result);
+    }
+  }
+
+  if (!doc.error) {
+    // add signature to doc
+    const fleetActionDoc = {
+      ...doc.EndpointActions,
+      '@timestamp': doc['@timestamp'],
+      agents,
+      timeout: 300, // 5 minutes
+      user_id: doc.user.id,
+    };
+    const fleetActionDocSignature = await endpointContext.service
+      .getMessageSigningService()
+      .sign(fleetActionDoc);
+    const signedFleetActionDoc = {
+      ...fleetActionDoc,
+      signed: {
+        data: fleetActionDocSignature.data.toString('base64'),
+        signature: fleetActionDocSignature.signature,
+      },
+    };
+    // write actions to .fleet-actions index
+    try {
+      const fleetActionIndexResult = await esClient.index<EndpointAction>(
+        {
+          index: AGENT_ACTIONS_INDEX,
+          body: signedFleetActionDoc,
+          refresh: 'wait_for',
+        },
+        {
+          meta: true,
+        }
+      );
+
+      if (fleetActionIndexResult.statusCode !== 201) {
+        throw new Error(fleetActionIndexResult.body.result);
+      }
+    } catch (e) {
+      // create entry in .logs-endpoint.action.responses-default data stream
+      // when writing to .fleet-actions fails
+      if (doesLogsEndpointActionsDsExist) {
+        await createFailedActionResponseEntry({
+          esClient,
+          doc: {
+            '@timestamp': moment().toISOString(),
+            agent: doc.agent,
+            EndpointActions: {
+              action_id: doc.EndpointActions.action_id,
+              completed_at: moment().toISOString(),
+              started_at: moment().toISOString(),
+              data: doc.EndpointActions.data,
+            },
+          },
+          logger,
+        });
+      }
+
+      throw e;
+    }
+  }
+};
+
+const createFailedActionResponseEntry = async ({
+  esClient,
+  doc,
+  logger,
+}: {
+  esClient: ElasticsearchClient;
+  doc: LogsEndpointActionResponse;
+  logger: Logger;
+}): Promise<void> => {
+  try {
+    await esClient.index<LogsEndpointActionResponse>({
+      index: `${ENDPOINT_ACTION_RESPONSES_DS}-default`,
+      body: {
+        ...doc,
+        error: {
+          code: failedFleetActionErrorCode,
+          message: 'Failed to deliver action request to fleet',
+        },
+      },
+    });
+  } catch (e) {
+    logger.error(e);
```

Comment:
Please confirm: this function will **always** return success even if an error was encountered, correct?

Also - suggestion: `wrapErrorIfNeeded()` on the error so that we get a better stack trace

Reply:
`wrapErrorIfNeeded` ae670fa8fb5a069970060ee2b77e893307f9cbb6